utils.go

```diff
@@ -14,14 +33,149 @@ type Storage struct {
 
 // String implements Storager.String
 func (s *Storage) String() string {
-	panic("implement me")
+	return fmt.Sprintf("Storager ftp {Name: %s, User: %s, WorkDir: %s}", s.name, s.user, s.workDir)
 }
 
 // NewStorager will create Storager only.
 func NewStorager(pairs ...types.Pair) (types.Storager, error) {
-	panic("implement me")
+	return newStoragerWithFTPClient(pairs...)
+}
+
+func newStoragerWithFTPClient(pairs ...types.Pair) (store *Storage, err error) {
+	defer func() {
+		if err != nil {
+			err = services.InitError{Op: "new_storager", Type: Type, Err: formatError(err), Pairs: pairs}
+		}
+	}()
+
+	store = &Storage{
+		connection: nil,
+		user:       "anonymous",
+		password:   "anonymous",
+		name:       "localhost:21",
+		workDir:    "/",
+	}
+
+	opt, err := parsePairStorageNew(pairs)
+	if err != nil {
+		return
+	}
+	if opt.HasName {
+		store.name = opt.Name
+	}
+	if opt.HasWorkDir {
+		store.workDir = opt.WorkDir
+	}
+	if opt.HasCredential {
+
+	}
+	err = store.connect()
+	if err != nil {
+		return nil, err
+	}
+	return
+}
+
+func (s *Storage) connect() error {
+	c, err := ftp.Dial(s.name, ftp.DialWithTimeout(5*time.Second))
+	if err != nil {
+		return err
+	}
+
+	err = c.Login(s.user, s.password)
+	if err != nil {
+		return err
+	}
+	err = c.ChangeDir(s.workDir)
+	if err != nil {
+		return err
+	}
+	s.connection = c
+	return err
+}
+
+func (s *Storage) makeDir(path string) (err error) {
+	rp := s.getAbsPath(path)
+	err = s.connection.MakeDir(rp)
+	return
+}
+
+func (s *Storage) getAbsPath(path string) string {
+	if filepath.IsAbs(path) {
+		return path
+	}
+	absPath := filepath.Join(s.workDir, path)
+
+	// Join will clean the trailing "/", we need to append it back.
+	if strings.HasSuffix(path, PathSeparator) {
+		absPath += PathSeparator
+	}
+	return absPath
+}
+
+func (s *Storage) getNameList(path string) (namelist []string, err error) {
+	namelist, err = s.connection.NameList(s.getAbsPath(path))
+	if err != nil {
+		return nil, err
+	}
+	return
+}
+
+func (s *Storage) newObject(done bool) *types.Object {
+	return types.NewObject(s, done)
+}
+
+func (s *Storage) mapMode(fet ftp.EntryType) types.ObjectMode {
+	switch fet {
+	case ftp.EntryTypeFile:
+		return types.ModeRead
+	case ftp.EntryTypeFolder:
+		return types.ModeDir
+	case ftp.EntryTypeLink:
+		return types.ModeLink
+	}
+	return types.ModeRead
+}
+
+func (s *Storage) formatFileObject(fe *ftp.Entry, parent string) (obj *types.Object, err error) {
+	obj = types.NewObject(s, false)
+	obj.ID = filepath.Join(parent, fe.Name)
+	obj.Mode = s.mapMode(fe.Type)
+	obj.Path = fe.Target
+	return
+}
+
+func formatError(err error) error {
+	if _, ok := err.(services.InternalError); ok {
+		return err
+	}
+	var estr string = err.Error()
+	ereg, regerr := regexp.Compile(estr)
```

Comment:
ftp returns `textproto.Error`, how about check error via `Code` instead of regexp?

For example:

https://github.com/rclone/rclone/blob/654575575810fb01f76e8b4ffb363adfe7473afe/backend/ftp/ftp.go#L250-L256